clients/src/main/java/org/apache/kafka/clients/admin/internals/PartitionLeaderStrategy.java

```diff
@@ -195,4 +201,93 @@ public LookupResult<TopicPartition> handleResponse(
         return new LookupResult<>(failed, mapped);
     }
 
+    /**
+     * This subclass of {@link AdminApiFuture} starts with a pre-fetched map for keys to broker ids which can be
+     * used to optimise the request. The map is kept up to date as metadata is fetching as this request is processed.
+     * This is useful for situations in which {@link PartitionLeaderStrategy} is used
+     * repeatedly, such as a sequence of identical calls to
+     * {@link org.apache.kafka.clients.admin.Admin#listOffsets(Map, org.apache.kafka.clients.admin.ListOffsetsOptions)}.
+     */
+    public static class PartitionLeaderFuture<V> implements AdminApiFuture<TopicPartition, V> {
+        private final Set<TopicPartition> requestKeys;
+        private final Map<TopicPartition, Integer> partitionLeaderCache;
+        private final Map<TopicPartition, KafkaFuture<V>> futures;
+
+        public PartitionLeaderFuture(Set<TopicPartition> requestKeys, Map<TopicPartition, Integer> partitionLeaderCache) {
+            this.requestKeys = requestKeys;
+            this.partitionLeaderCache = partitionLeaderCache;
+            this.futures = requestKeys.stream().collect(Collectors.toMap(
+                Function.identity(),
+                k -> new KafkaFutureImpl<>()
+            ));
+        }
+
+        @Override
+        public Set<TopicPartition> lookupKeys() {
+            return futures.keySet();
+        }
+
+        @Override
+        public Set<TopicPartition> uncachedLookupKeys() {
+            Set<TopicPartition> keys = new HashSet<>();
+            requestKeys.forEach(tp -> {
+                if (!partitionLeaderCache.containsKey(tp)) {
+                    keys.add(tp);
+                }
+            });
+            return keys;
+        }
+
+        @Override
+        public Map<TopicPartition, Integer> cachedKeyBrokerIdMapping() {
+            Map<TopicPartition, Integer> mapping = new HashMap<>();
+            partitionLeaderCache.forEach((tp, brokerId) -> {
```

Comment:
The size of `partitionLeaderCache` is larger than `requestKeys`, correct? If that's the case, we should loop over `requestKeys` instead of `partitionLeaderCache`.